### ISSUE TITLE
Setup Exact Target for MoFo transactional emails

### ISSUE THREAD

**Issue by adamlofting** [MozillaFoundation/plan#433]
Following the discussion and vendor review in issue #242, we need to setup Exact Target so that we can use it for transactional emails with any of our software products. The immediate task is one of 'platform engineering' before we can work on content and user journeys. I've tried to keep this ticket focused on that.

### Audience
The result of this implementation project will be used by:

* MoFo Engineers: implementing transactional emails in our products
* MoFo Comms / Product Staff: writing editing and testing transactional emails
* MoFo Metrics: understanding what's working

### Success
* MoFo engineers have knowledge and documentation about how to connect a product to ExactTarget, and how to trigger a transactional email in code
* MoFo staff who write emails know how to setup a new email, edit and existing email, and test email content, and how to work with Engineers to get these emails triggered in code
* MoFo metrics: happy that the UX is properly instrumented and that we can report on the data we need to optimize this UX over time

### Vision
1. Training for a selection of engineers and staff to scope the implementation project 
2. Proof of concept implementation with documentation so other engineers without training can replicate this **(Current stage)**
3. Implementation on teach.mozilla.org
4. Reporting setup for teach.m.o transactional emails
5. Tickets filed and scheduled to replace all existing transactional emails in MoFo products that will be maintained going forward

### Measurement
For this implementation task, completing the five steps outlined above will be our measure of success. There will be iterative future work (and tickets to be filed) to optimize our email messaging and effectiveness in relation to our Retention KPIs.

### Roles
 * Phase: Initial Tech Implementation
 * Driver: @adamlofting
 * Decision Maker: @simonwex
 * Design Lead: n/a
 * Development Lead: TBC
 * Quality Verifier: @thisandagain @hannahkane @HPaulJohnson 
 * Consulted: @valianttry

**Comment by adamlofting** [comment]
Via MoCo, we are currently scheduling ExactTarget training to support this task. Expected availability is for early May, though @simonwex and I are to meet with DEG (the training company) sooner to scope our training needs. They have been pointed at this ticket as a starting point.

In the meantime we should explore our ExactTarget sandbox account, and the product documentation to see how far we can get ourselves.

**Comment by adamlofting** [comment]
@simonwex at some point in the next few weeks, this will be ready for more serious dev input. We'll need to think about resourcing for it then.

**Comment by adamlofting** [comment]
Initial call is happening today/tonight (depending on your timezone). Updates to follow after that.

**Comment by adamlofting** [comment]
DEG are providing training next week for dev team: @alicoding @mmmavis  @simonwex @ScottDowne (appointments are due soon if they haven't already been sent).

Please add specific questions/requirements to this Etherpad before the training:
https://etherpad.mozilla.org/exacttarget-training-questions

Training will be recorded. :video_camera:

**Comment by adamlofting** [comment]
This work is ongoing. Training is happening in this Heartbeat, but let's consider making it a P1 in the following heartbeat (May 22) for implementation.

**Comment by OpenMatt** [comment]
Andrew: would like to be able to sandbox and test by end of May

**Comment by jbuck** [comment]
@adamlofting Mind if I join that?

**Comment by adamlofting** [comment]
@jbuck, absolutely I'll forward some emails to you now

**Comment by adamlofting** [comment]
Training happened last week. This now needs developer time assigned to complete the next steps. 

- [ ] Proof of concept implementation with documentation so other engineers without training can replicate this
- [ ] Implementation on teach.mozilla.org

Flag @simonwex for the next heartbeat scheduling.

**Comment by OpenMatt** [comment]
@simonwex should we file a new ticket for the dev work? or just roll this one to May 22 heartbeat?

**Comment by adamlofting** [comment]
I'd like to keep this as the running `plan` ticket, and file any smaller tickets in [mofo-crm](https://github.com/MozillaFoundation/mofo-crm/issues) please. After this next dev work there will be reporting and pm tasks to wrap this up.

**Comment by OpenMatt** [comment]
note: significant lift for dev on this going forward

**Comment by adamlofting** [comment]
Training didn't cover the requirements we asked for and dev lift could be heavy based on Andrew's experimentation.

**Comment by OpenMatt** [comment]
* may also need @ScottDowne : Erika and Simon to co-ordinate

**Comment by edrushka** [comment]
flagging so I can follow

**Comment by adamlofting** [comment]
@alicoding is going to spend some time on this this heartbeat. We just met to kick this off, and the current plan is:

1. Proof of concept command line script that can trigger and email via ExactTarget's API
2. Turn into a consumable module we can use in other projects

Documentation for reference is here:
http://help.exacttarget.com/en/documentation/exacttarget/interactions/triggered_emails/

I've requested an account for Ali in the ET Sandbox environment.

**Comment by alicoding** [comment]
:+1:

**Comment by simonwex** [comment]
As @thisandagain discovered, the API is horrible. I'm chatting with @pmclanahan tomorrow to see if he can make our lives easier by adding some of this to https://github.com/mozilla/basket

**Comment by adamlofting** [comment]
@simonwex excellent! 

--

For this thread in general, just in case it's a useful reference, there is a [jsforce](https://github.com/jsforce/jsforce) library. It unfortunately doesn't talk to the ET API (at least not yet), but it does have some example code for talking to other Salesforce SOAP APIs ([e.g](https://github.com/jsforce/jsforce/blob/master/lib/soap.js).).

**Comment by simonwex** [comment]
@jbuck and I had a lovely chat with @pmclanahan. He agrees there's benefit to adding to the Basket API to support transactional emails. Basket currently makes _use_ of some of the transactional ET API calls but needs to be exposed through the external Basket API. -- so some of this pain has already been suffered. We thank Paul for his bravery.

Next steps are for @jbuck to spec a plan for implementation in the basket repo. https://github.com/mozilla/basket

**Comment by davidascher** [comment]
awesome.

**Comment by jbuck** [comment]
https://gist.github.com/thisandagain/36be0cba730729feba26

**Comment by adamlofting** [comment]
:sunny:

**Comment by adamlofting** [comment]
While we're thinking through how to add triggered emails to Basket, would it be a good use of @alicoding's time to start writing a node wrapper/module that let's us talk to basket?

For example, when our email lists are moved to ET, we're going to want to use the subscribe/unsubscribe end points so working out how to talk to these now will also get us ready for talking to the triggered email endpoints when they are ready.

**Comment by pmclanahan** [comment]
They got a start on that in gaia: https://github.com/mozilla-b2g/gaia/blob/v1.4/apps/communications/ftu/js/basket-client.js

**Comment by adamlofting** [comment]
@jbuck / @alicoding - could I get a status update on this please? Have we been able to make any progress?

**Comment by alicoding** [comment]
So, last week I've spent most of my time reading their documentations and
haven't got the account to explore deeper on ET part, yet. I'm also got
pulled off to work on advocacy stuff since I'll be away next week for
JSConf, so I don't have any progress on my part yet.

**Comment by edrushka** [comment]
Do we need to create accounts for this?

**Comment by jbuck** [comment]
Yes, you'll probably need to get an account from jessilyn

**Comment by OpenMatt** [comment]
status?

**Comment by adamlofting** [comment]
This didn't get the dev time it needed. I'd like to talk about this one on the planning call to see what priority it should have in the upcoming heartbeat(s).

**Comment by OpenMatt** [comment]
(via @thisandagain in today's Heartbeat call:) 
Product dependencies / implications: we can continue to use Sawmill, so this ticket is not a ship blocker. BUT: 
* This means no localized transactional email. 
* And: no push notifications, which will impact marketing and retention.

**Comment by adamlofting** [comment]
@simonwex and I just met to talk about next steps.

Goal is for @jbuck to move this forward in the current heartbeat, with all the things we can work on right away. I'll book time to talk with @jbuck tomorrow but I think this will be:

1. Planning how we talk to Basket https://github.com/MozillaFoundation/mofo-devops/issues/73 and talking to @pmclanahan about what changes we need to make, and scheduling this in
2. Getting started on the node-module part of this that we will re-use across our apps (setting up the repo, docs, tests, etc)

**Comment by simonwex** [comment]
@adamlofting with you as the driver of this, it should be assigned to you. I suggest filing a separate issue in the devlops repo to make sure it surfaces properly on @jbuck's desk.

**Comment by adamlofting** [comment]
Yep, the separate issue is filed here: 

* https://github.com/MozillaFoundation/mofo-devops/issues/73

**Comment by adamlofting** [comment]
Well, this was certainly a painful ticket :see_no_evil:

Similar requirements still exist, but I'm not sure this ticket is helpful right now.
